domain/fhir/fhir_repository.go

```diff
@@ -0,0 +1,103 @@
+package fhir
+
+import (
+	"context"
+	"fmt"
+	"github.com/google/uuid"
+	"github.com/tidwall/gjson"
+	"strings"
+	"time"
+)
+
+type fhirTask struct {
+	data gjson.Result
+}
+
+func (task fhirTask) MarshalToTask() (*Task, error) {
+	if rType := task.data.Get("resourceType").String(); rType != "Task" {
+		return nil, fmt.Errorf("invalid resource type. got: %s, expected Task", rType)
+	}
+	codeQuery := fmt.Sprintf("code.coding.#(system==%s).code", SnomedCodingSystem)
+	if codeValue := task.data.Get(codeQuery).String(); codeValue != string(SnomedTransferCode) {
+		return nil, fmt.Errorf("unexpecting coding: %s", codeValue)
+	}
+	patientID := ""
+	if parts := strings.Split(task.data.Get("for.reference").String(), "/"); len(parts) > 1 {
+		patientID = parts[1]
+	}
+	return &Task{
+		ID: task.data.Get("id").String(),
+		TaskProperties: TaskProperties{
+			Status:      task.data.Get("status").String(),
+			OwnerID:     task.data.Get("owner.identifier.value").String(),
+			RequesterID: task.data.Get("requester.agent.identifier.value").String(),
+			PatientID:   patientID,
+		},
+	}, nil
+}
+
+type fhirRepository struct {
+	client      Client
+	taskFactory TaskFactory
+}
+
+func NewFHIRRepository(fhirClient Client) *fhirRepository {
+	return &fhirRepository{
+		client:      fhirClient,
+		taskFactory: TaskFactory{},
+	}
+}
+
+func (r fhirRepository) CreateTask(ctx context.Context, taskProperties TaskProperties) (*Task, error) {
+	task := r.taskFactory.New(taskProperties)
+
+	fhirData := map[string]interface{}{
+		"resourceType": "Task",
+		"id":           task.ID,
+		"status":       task.Status,
+		// TODO: patient seems mandatory in the spec, but can only be sent when placer already
+		// has patient in care to protect the identity of the patient during the negotiation phase.
+		//"for": map[string]string{
+		//	"reference": fmt.Sprintf("Patient/%s", domainTask.PatientID),
+		//},
```

Comment:
This doesn't really matter right? As we can simply include the patient ID here. The fact that the other system won't be able to get the patient details until proper authentication has taken place is not relevant right?

Reply:
This is existing code (looks new due to refactoring) from @stevenvegt so I'd like to discuss that with him what we do with it.